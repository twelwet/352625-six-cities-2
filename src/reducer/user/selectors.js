import NameSpace from "../name-spaces";

const NAME_SPACE = NameSpace.USER;

const getAuthFlag = (state) => {
  return state[NAME_SPACE].isAuthRequired;
};

const getEmail = (state) => {
  return state[NAME_SPACE].email;
};

export {getAuthFlag, getEmail};

