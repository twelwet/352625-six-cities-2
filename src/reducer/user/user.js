const initialState = {
  isAuthRequired: true,
  id: null,
  email: null,
  name: null,
  [`avatar_url`]: null,
  [`is_pro`]: false,
  isError: false,
  errorType: null
};

const ActionType = {
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  SIGN_IN: `SIGN_IN`,
  SIGN_IN_FAIL: `SIGN_IN_FAIL`
};

const ActionCreator = {
  requireAuthorization: (status = initialState.isAuthRequired) => {
    return {
      type: ActionType.REQUIRE_AUTHORIZATION,
      payload: status
    };
  },
  signIn: (userData) => {
    return {
      type: ActionType.SIGN_IN,
      error: false,
      payload: userData
    };
  },
  signInFail: (error) => {
    return {
      type: ActionType.SIGN_IN_FAIL,
      error: true,
      payload: error
    };
  }
};

const Operation = {
  signIn: () => (dispatch, _, api) => {
    return api.post(`/login`, {
      email: `user@gmail.com`,
      password: `password`
    })
      .then((response) => {
        dispatch(ActionCreator.signIn(response.data));
        dispatch(ActionCreator.requireAuthorization(false));
      })
      .catch((err) => {
        dispatch(ActionCreator.signInFail(err));
        dispatch(ActionCreator.requireAuthorization(true));
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRE_AUTHORIZATION:
      return Object.assign({}, state, {
        isAuthRequired: action.payload
      });
    case ActionType.SIGN_IN:
      return Object.assign({}, state, {
        id: action.payload.id,
        email: action.payload.email,
        name: action.payload.name,
        [`avatar_url`]: action.payload[`avatar_url`],
        [`is_pro`]: action.payload[`is_pro`],
        isError: action.error,
        errorType: null
      });
    case ActionType.SIGN_IN_FAIL:
      return Object.assign({}, state, {
        isError: action.error,
        errorType: action.payload.message
      });
  }

  return Object.assign({}, state);
};

export {ActionType, ActionCreator, reducer};
