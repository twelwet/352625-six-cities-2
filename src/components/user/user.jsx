import React from "react";
import PropTypes from "prop-types";
import {getAuthFlag, getEmail} from "../../reducer/user/selectors";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

const User = (props) => {
  const {isAuthRequired, email} = props;
  return (
    <li className="header__nav-item user">
      <Link className="header__nav-link header__nav-link--profile"
        to={isAuthRequired ? `/login` : `/favorites`}>
        <div className="header__avatar-wrapper user__avatar-wrapper">
        </div>
        <span className="header__user-name user__name">{isAuthRequired ? `Sign In` : email}</span>
      </Link>
    </li>
  );
};

User.propTypes = {
  isAuthRequired: PropTypes.bool.isRequired,
  email: PropTypes.string
};

const mapStateToProps = (state) => ({
  isAuthRequired: getAuthFlag(state),
  email: getEmail(state)
});

export {User};

export default connect(mapStateToProps)(User);
