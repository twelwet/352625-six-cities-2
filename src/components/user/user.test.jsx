import React from "react";
import renderer from "react-test-renderer";
import {User} from "./user.jsx";
import {BrowserRouter} from "react-router-dom";

const props = {
  isAuthRequired: true,
  email: `user@gmail.com`
};

it(`User correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<BrowserRouter>
      <User {...props} />
    </BrowserRouter>);

  expect(tree).toMatchSnapshot();
});
