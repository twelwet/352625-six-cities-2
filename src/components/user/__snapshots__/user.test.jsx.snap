// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User correctly renders after relaunch 1`] = `
<li
  className="header__nav-item user"
>
  <a
    className="header__nav-link header__nav-link--profile"
    href="/login"
    onClick={[Function]}
  >
    <div
      className="header__avatar-wrapper user__avatar-wrapper"
    />
    <span
      className="header__user-name user__name"
    >
      Sign In
    </span>
  </a>
</li>
`;
